src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,157 @@
+        Node<T> newNode = new Node<>(tail, null, value);
+        if (head == null) {
+            head = newNode;
+        }
+        if (tail != null) {
+            tail.next = newNode;
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size()) {
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, head, value);
+            head.prev = newNode;
+            head = newNode;
+        } else {
+            Node<T> currentNode = getNode(index);
+            Node<T> newNode = new Node<>(currentNode.prev, currentNode, value);
+            currentNode.prev.next = newNode;
+            currentNode.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> pointer = getNode(index);
+        if (pointer == null) {
+            return null;
+        } else {
+            return pointer.value;
+        }
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> pointer = getNode(index);
+        T oldValue = null;
+        if (pointer != null) {
+            oldValue = pointer.value;
+            pointer.value = value;
+        }
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
+        checkIndex(index);
+        //   T toReturn = null;
+        Node<T> pointer = getNode(index);
+        if (pointer != null) {
+            T result = pointer.value;
+            unlink(pointer);
+            return result;
+        }
         return null;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> pointer = head;
+        while (pointer != null) {
+            if (pointer.value == object || pointer.value != null
+                    && (pointer.value.equals(object))) {
+                unlink(pointer);
+                return true;
+            }
+            pointer = pointer.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void unlink(Node<T> node) {
+        if (node.prev == null) {
+            head = node.next;
+        } else {
+            node.prev.next = node.next;
+        }
+        if (node.next == null) {
+            tail = node.prev;
+        } else {
+            node.next.prev = node.prev;
+        }
+        size--;
+    }
+
+    private Node<T> getNode(int index) {
+        int count = 0;
+        Node<T> pointer = head;
+        while (pointer != null) {
```

Comment:
use for loop instead of while with the counter. Also, check which path is shorter - from the head or from the tail, and iterate from there, this will boost your performance